lib/mosaic-routing/src/main/java/org/eclipse/mosaic/lib/routing/graphhopper/GraphHopperRouting.java

```diff
@@ -127,69 +135,100 @@ public GraphHopperRouting loadGraphFromDatabase(Database db) {
+        final TurnWeighting weighting = new TurnWeightingOptional(
+                graphhopperWeighting, (TurnCostExtension) queryGraph.getExtension(),
+                routingRequest.getRoutingParameters().isConsiderTurnCosts(),
+                routingRequest.getRoutingParameters().getRestrictionCosts()
+        );
 
         // create algorithm
-        AlternativeRoutesRoutingAlgorithm algo = new DijkstraCamvitChoiceRouting(queryGraph, weighting);
+        final AlternativeRoutesRoutingAlgorithm algo = new DijkstraCamvitChoiceRouting(queryGraph, weighting);
         algo.setRequestAlternatives(routingRequest.getRoutingParameters().getNumAlternativeRoutes());
 
+        final List<Path> paths = new ArrayList<>();
+
         // Calculates all paths and returns the best one
-        paths.add(algo.calcPath(qrFrom.getClosestNode(), qrTo.getClosestNode()));
+        paths.add(algo.calcPath(querySource.getClosestNode(), queryTarget.getClosestNode()));
 
         //add alternative paths to path set
         paths.addAll(algo.getAlternativePaths());
 
         final Set<String> duplicateSet = new HashSet<>();
+        final List<CandidateRoute> result = new ArrayList<>();
 
         // convert paths to routes
-        for (Path path : paths) {
-            final CandidateRoute route = preparePath(path, queryGraph, qrFrom, qrTo, target);
+        for (final Path path : paths) {
+            final CandidateRoute route = preparePath(path, queryGraph, querySource, queryTarget, target);
             if (route != null
                     && !route.getConnectionIds().isEmpty()
                     && checkForDuplicate(route, duplicateSet)
                     && checkRouteOnRequiredSourceConnection(route, source)) {
-                routesList.add(route);
+                result.add(route);
             }
         }
-        return routesList;
+        return result;
+    }
+
+    private QueryResult createQueryForTarget(FlagEncoder flagEncoder, RoutingPosition target) {
+        final EdgeFilter toEdgeFilter = createEdgeFilterForRoutingPosition(target, flagEncoder);
+        QueryResult qrTo = ghApi.getLocationIndex().findClosest(target.getPosition().getLatitude(), target.getPosition().getLongitude(), toEdgeFilter);
+        if (target.getConnectionId() != null) {
+            return fixQueryResultIfNoClosestEdgeFound(target, qrTo, flagEncoder);
+        } else {
+            return fixQueryResultIfSnappedPointIsCloseToTowerNode(qrTo, target);
+        }
+    }
+
+    private QueryResult createQueryForSource(FlagEncoder flagEncoder, RoutingPosition source) {
+        final EdgeFilter fromEdgeFilter = createEdgeFilterForRoutingPosition(source, flagEncoder);
+        QueryResult qrFrom = ghApi.getLocationIndex().findClosest(source.getPosition().getLatitude(), source.getPosition().getLongitude(), fromEdgeFilter);
+        if (source.getConnectionId() != null) {
+            qrFrom = fixQueryResultIfSnappedPointIsTowerNode(qrFrom, source, fromEdgeFilter);
```

Comment:
As far as I see this is the only difference from `creqteQueryForTarget`, should we unify the two methods?
Also is there a reason why we wouldn't snap the target position to a Tower Node?

Reply:
This is a special treatment for the source when we are doing route request from applications, as we **always** require to match the edge the vehicle is currently on. Otherwise the calculated routes can not be used with SUMO. For the target this is not explicitly required. This is also related to your other comment: we want to find an edge, and it is much more likely to match an edge if we search with a position which is located ON the connection, and not at the beginning of it (as it was before). This was a leftover when we unified edges and connections.

Tl;dr: we could also do this for the target, but I cannot estimate the implications of this right now.